### PR TITLE
[feature](nereids)support set command

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes
**test case:**
setPassword and setUserProperties:
regression-test/suites/account_p0/test_auth_compatibility.groovy

setDefaultStorageVault:
regression-test/suites/vaults/privilege.groovy

setUserVariable:
regression-test/suites/query_p0/set/test_user_var.groovy

setLdapAdminPassword:
regression-test/suites/auth_p0/test_set_ldap_admin_password_auth.groovy

other commands:
regression-test/suites/query_p0/set/test_set_command.groovy

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/AuditLogHelper.java [hunk 1]
```diff
@@ -259,10 +260,20 @@ private static void logAuditLogImpl(ConnectContext ctx, String origStmt, Stateme
         auditEventBuilder.setFeIp(FrontendOptions.getLocalHostAddress());
 
         // We put origin query stmt at the end of audit log, for parsing the log more convenient.
-        if (!ctx.getState().isQuery() && (parsedStmt != null && parsedStmt.needAuditEncryption())) {
-            auditEventBuilder.setStmt(parsedStmt.toSql());
+        if (parsedStmt instanceof LogicalPlanAdapter) {
+            if (!ctx.getState().isQuery() && (parsedStmt != null
+                    && (((LogicalPlanAdapter) parsedStmt).getLogicalPlan() instanceof Command))
+                    && ((Command) ((LogicalPlanAdapter) parsedStmt).getLogicalPlan()).needAuditEncryption()) {
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/qe/AuditLogHelper.java:266]: use trait here toSql and needAuditEncryption() must be implemented together

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/SetSessionVarOp.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        sb.append(" ").append(name).append(" = ").append(value.toSql());
+        return sb.toString();
+    }
+
+    public void forwardToMaster(ConnectContext ctx) throws Exception {
+        setType(SetType.SESSION);
+        VariableMgr.setVarForNonMasterFE(ctx.getSessionVariable(), translateToLegacyVar(ctx));
+    }
+
+    private SetVar translateToLegacyVar(ConnectContext ctx) {
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/SetSessionVarOp.java:135]: todo remove it in next pr

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java [hunk 3]
```diff
@@ -116,4 +116,14 @@ public String treeString() {
...
+
+    public String toSql() {
+        return "";
+    }
+
+    public boolean needAuditEncryption() {
+        return false;
+    }
+
+    public void forwardToMaster(ConnectContext ctx) throws Exception {}
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java:128]: change name to afterForwardToMaster

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/SetOptionsCommand.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.doris.nereids.trees.plans.visitor.PlanVisitor;
+import org.apache.doris.qe.ConnectContext;
+import org.apache.doris.qe.StmtExecutor;
+
+import java.util.List;
+
+/**
+ * SetOptionsCommand
+ */
+public class SetOptionsCommand extends Command implements ForwardWithSync {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/SetOptionsCommand.java:36]: in `StmtExecutor#L707`, we use `instance of Forward` to decide whether to forward to master. we should use toRedirectStatus result decide it.

**Review by morrySnow — COMMENTED** [review]
add test for each set statement

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 5]
```diff
@@ -3829,4 +3855,113 @@ public LogicalPlan visitShowConfig(ShowConfigContext ctx) {
...
+    public SetVarOp visitSetCollate(SetCollateContext ctx) {
+        String charset = ctx.charsetName != null ? stripQuotes(ctx.charsetName.getText()) : null;
+        String collate = ctx.collateName != null ? stripQuotes(ctx.collateName.getText()) : null;
+        return new SetCharsetAndCollateVarOp(charset, collate);
+    }
+
+    @Override
+    public SetVarOp visitSetLdapAdminPassword(SetLdapAdminPasswordContext ctx) {
+        String passwordText = stripQuotes(ctx.STRING_LITERAL().getText());
+        boolean isPlain = ctx.LEFT_PAREN() != null;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:3936]:
```suggestion
        boolean isPlain = ctx.PASSWORD() != null;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java [hunk 6]
```diff
@@ -116,4 +116,14 @@ public String treeString() {
...
         throw new RuntimeException("Command do not implement withGroupExpression");
     }
+
+    public String toSql() {
+        return "";
+    }
+
+    public boolean needAuditEncryption() {
+        return false;
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java:126]: command should not have these interface

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java [hunk 7]
```diff
@@ -116,4 +116,14 @@ public String treeString() {
...
+
+    public void afterForwardToMaster(ConnectContext ctx) throws Exception {}
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/Command.java:128]: maybe this interface should in `Forward` trait

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 8]
```diff
@@ -820,7 +820,7 @@ optionWithoutType
     | NAMES (charsetName=identifierOrText | DEFAULT)
         (COLLATE collateName=identifierOrText | DEFAULT)?               #setCollate
     | PASSWORD (FOR userIdentify)? EQ (STRING_LITERAL
-        | (PASSWORD LEFT_PAREN STRING_LITERAL RIGHT_PAREN))             #setPassword
+        | (isPlain=PASSWORD LEFT_PAREN STRING_LITERAL RIGHT_PAREN))             #setPassword
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:823]: maybe should be `isNotPlain` ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 9]
```diff
@@ -3829,4 +3855,111 @@ public LogicalPlan visitShowConfig(ShowConfigContext ctx) {
...
+        String passwordText;
+        UserIdentity userIdentity = null;
+        if (ctx.userIdentify() != null) {
+            user = stripQuotes(ctx.userIdentify().user.getText());
+            host = ctx.userIdentify().host != null ? stripQuotes(ctx.userIdentify().host.getText()) : "%";
+            isDomain = ctx.userIdentify().ATSIGN() != null;
+            userIdentity = new UserIdentity(user, host, isDomain);
+        }
+        passwordText = stripQuotes(ctx.STRING_LITERAL().getText());
+        return new SetPassVarOp(userIdentity, new PassVar(passwordText, ctx.isPlain != null));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:3910]:
```suggestion
        return new SetPassVarOp(userIdentity, new PassVar(passwordText, ctx.isNotPlain == null));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 10]
```diff
@@ -3829,4 +3855,111 @@ public LogicalPlan visitShowConfig(ShowConfigContext ctx) {
...
+    public SetVarOp visitSetCollate(SetCollateContext ctx) {
+        String charset = ctx.charsetName != null ? stripQuotes(ctx.charsetName.getText()) : null;
+        String collate = ctx.collateName != null ? stripQuotes(ctx.collateName.getText()) : null;
+        return new SetCharsetAndCollateVarOp(charset, collate);
+    }
+
+    @Override
+    public SetVarOp visitSetLdapAdminPassword(SetLdapAdminPasswordContext ctx) {
+        String passwordText = stripQuotes(ctx.STRING_LITERAL().getText());
+        boolean isPlain = ctx.PASSWORD() != null;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:3934]:
```suggestion
        boolean isPlain = ctx.PASSWORD() == null;
```

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**